File: backend/app/tasks/grading.py
"""Celery tasks for AI grading"""
from celery import Task
from celery.utils.log import get_task_logger
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from datetime import datetime, timedelta
from typing import Dict, Any
import asyncio

from app.core.celery_app import celery_app
from app.db.session import AsyncSessionLocal
from app.models.submission import Submission, SubmissionStatus, ExamSession, SessionStatus
from app.models.exam import Exam, TaskType, ExamStatus
from app.services.ai_grading import grade_submission
from app.services.storage import storage_service

logger = get_task_logger(__name__)


class DatabaseTask(Task):
    """Base task with database session handling"""
    
    _db = None
    
    def after_return(self, *args, **kwargs):
        """Close database session after task completes"""
        if self._db is not None:
            self._db.close()


@celery_app.task(
    bind=True,
    base=DatabaseTask,
    name="app.tasks.grading.grade_submission_task",
    max_retries=3,
    default_retry_delay=120,  # 2 minutes
    acks_late=True
)
def grade_submission_task(self, submission_id: str):
    """
    Async task to grade a submission using AI
    
    Args:
        submission_id: Submission ID to grade
    """
    logger.info(f"Starting AI grading task for submission {submission_id}")
    
    # Run async code in event loop
    loop = asyncio.get_event_loop()
    result = loop.run_until_complete(_grade_submission_async(submission_id, self))
    
    return result


async def _grade_submission_async(submission_id: str, task: Task) -> Dict[str, Any]:
    """Async implementation of grading task"""
    async with AsyncSessionLocal() as db:
        try:
            # Get submission with all relationships
            result = await db.execute(
                select(Submission)
                .where(Submission.id == submission_id)
                .options(
                    selectinload(Submission.images),
                    selectinload(Submission.session).selectinload(ExamSession.exam)
                    .selectinload(Exam.task_types).selectinload(TaskType.variants)
                )
            )
            submission = result.scalar_one_or_none()
            
            if not submission:
                logger.error(f"Submission {submission_id} not found")
                return {"error": "Submission not found"}
            
            # Check if already processed
            if submission.status not in [SubmissionStatus.PROCESSING, SubmissionStatus.FLAGGED]:
                logger.info(f"Submission {submission_id} already processed with status {submission.status}")
                return {"status": "already_processed", "submission_status": submission.status}
            
            # Update status
            submission.status = SubmissionStatus.PROCESSING
            submission.ai_request_started_at = datetime.utcnow()
            await db.commit()
            
            # Get session and exam
            session = submission.session
            exam = session.exam
            
            # Get image URLs
            image_urls = []
            for image in sorted(submission.images, key=lambda x: x.order_index):
                if image.file_path.startswith("submissions/"):
                    # S3 storage - generate presigned URL
                    try:
                        presigned_url = await storage_service.generate_presigned_url(
                            s3_key=image.file_path,
                            expires_in=3600  # 1 hour for AI processing
                        )
                        image_urls.append(presigned_url)
                    except Exception as e:
                        logger.error(f"Failed to generate presigned URL for image {image.id}: {e}")
                        raise
                else:
                    # Legacy local storage - should not happen in production
                    logger.error(f"Image {image.id} is stored in local storage (not S3). Skipping AI processing.")
                    raise ValueError(f"Image {image.id} is not stored in S3. All images must be in S3 for AI processing.")
            
            if not image_urls:
                raise ValueError("No accessible images found for AI processing")
            
            # Build task description from variants
            task_descriptions = []
            reference_solutions = []
            rubric_items = []
            total_max_score = 0
            
            for task_type in exam.task_types:
                variant_id = session.variant_assignments.get(task_type.id)
                variant = next((v for v in task_type.variants if v.id == variant_id), None)
                
                if variant:
                    task_descriptions.append(f"""
Задача {task_type.order_index + 1}: {task_type.title}
{task_type.description}

Вариант:
{variant.content}
""")
                    
                    # Add reference solution if available
                    if variant.reference_solution:
                        reference_solutions.append(f"""
Эталонное решение для задачи {task_type.order_index + 1}:
{variant.reference_solution}
""")
                    
                    rubric_items.append({
                        "task_type": task_type.title,
                        "max_score": task_type.max_score,
                        "criteria": "Оценивать по критериям в системном промпте"
                    })
                    total_max_score += task_type.max_score
            
            task_description = "\n".join(task_descriptions)
            reference_solution_text = "\n".join(reference_solutions) if reference_solutions else "См. критерии оценивания"
            rubric = {"criteria": rubric_items, "total_max_score": total_max_score}
            
            # Call AI grading service
            logger.info(f"Calling AI grading for submission {submission_id}")
            ai_result = await grade_submission(
                images=image_urls,
                task_description=task_description,
                reference_solution=reference_solution_text,
                rubric=rubric,
                max_score=total_max_score,
                chemistry_rules={},
                submission_id=submission_id
            )
            
            # Record completion time
            submission.ai_request_completed_at = datetime.utcnow()
            
            # Extract metadata
            metadata = ai_result.pop("_metadata", {})
            if metadata:
                submission.ai_request_duration_seconds = metadata.get("duration_seconds")
            
            # Update submission based on result
            if "error" in ai_result:
                # AI processing failed - retry
                submission.status = SubmissionStatus.FLAGGED
                submission.ai_error = ai_result["error"]
                submission.flag_reasons = ["ai_processing_error"]
                logger.error(f"AI grading failed for submission {submission_id}: {ai_result['error']}")
                
                # Retry if not exceeded max retries
                if submission.ai_retry_count < 3:
                    submission.ai_retry_count += 1
                    await db.commit()
                    raise task.retry(countdown=120)  # Retry in 2 minutes
                
            elif ai_result.get("unreadable"):
                # Images are unreadable
                submission.status = SubmissionStatus.FLAGGED
                submission.ai_error = ai_result.get("unreadable_reason")
                submission.flag_reasons = ["unreadable_images"]
                logger.warning(f"Submission {submission_id} marked as unreadable: {ai_result.get('unreadable_reason')}")
            else:
                # Success
                submission.status = SubmissionStatus.PRELIMINARY
                submission.ai_score = ai_result.get("total_score")
                submission.ai_analysis = ai_result
                submission.ai_comments = ai_result.get("feedback")
                submission.ai_processed_at = datetime.utcnow()

                # Store transcriptions if present
                per_page = ai_result.get("per_page_transcriptions")
                if per_page:
                    # Map by order_index
                    sorted_images = sorted(submission.images, key=lambda x: x.order_index)
                    for idx, image in enumerate(sorted_images):
                        if idx < len(per_page):
                            image.ocr_text = per_page[idx]
                            image.processed_at = datetime.utcnow()
                logger.info(f"AI grading successful for submission {submission_id}: {submission.ai_score}/{total_max_score}")
            
            await db.commit()
            
            return {
                "status": "success",
                "submission_id": submission_id,
                "submission_status": submission.status,
                "ai_score": submission.ai_score,
                "max_score": total_max_score
            }
            
        except Exception as e:
            logger.error(f"Error in grading task for submission {submission_id}: {e}", exc_info=True)
            
            # Update submission with error
            try:
                result = await db.execute(
                    select(Submission).where(Submission.id == submission_id)
                )
                submission = result.scalar_one_or_none()
                if submission:
                    submission.status = SubmissionStatus.FLAGGED
                    submission.ai_error = str(e)
                    submission.flag_reasons = ["task_error"]
                    submission.ai_retry_count = (submission.ai_retry_count or 0) + 1
                    await db.commit()
            except Exception as commit_error:
                logger.error(f"Failed to update submission after error: {commit_error}")
            
            # Retry if possible
            if task.request.retries < task.max_retries:
                raise task.retry(exc=e, countdown=120)
            
            return {"error": str(e), "submission_id": submission_id}


@celery_app.task(name="app.tasks.grading.process_completed_exams")
def process_completed_exams():
    """
    Scheduled task to process submissions from completed exams
    Runs every 5 minutes
    """
    logger.info("Starting scheduled task: process_completed_exams")
    
    loop = asyncio.get_event_loop()
    result = loop.run_until_complete(_process_completed_exams_async())
    
    return result


async def _process_completed_exams_async():
    """Find and queue submissions from completed exams"""
    async with AsyncSessionLocal() as db:
        try:
            now = datetime.utcnow()
            
            # Find exams that ended in the last hour and are still active
            result = await db.execute(
                select(Exam)
                .where(
                    Exam.status.in_([ExamStatus.ACTIVE, ExamStatus.PUBLISHED]),
                    Exam.end_time <= now,
                    Exam.end_time >= now - timedelta(hours=1)
                )
            )
            completed_exams = result.scalars().all()
            
            if not completed_exams:
                logger.info("No recently completed exams found")
                return {"processed_exams": 0, "queued_submissions": 0}
            
            logger.info(f"Found {len(completed_exams)} recently completed exams")
            
            queued_count = 0
            for exam in completed_exams:
                # Find unprocessed submissions for this exam
                submissions_result = await db.execute(
                    select(Submission)
                    .join(ExamSession)
                    .where(
                        ExamSession.exam_id == exam.id,
                        Submission.status == SubmissionStatus.UPLOADED
                    )
                    .options(selectinload(Submission.images))
                )
                submissions = submissions_result.scalars().all()
                
                logger.info(f"Exam {exam.id} ({exam.title}): found {len(submissions)} unprocessed submissions")
                
                # Queue each submission for grading
                for submission in submissions:
                    # Skip empty submissions (no images)
                    if not submission.images or len(submission.images) == 0:
                        continue
                    try:
                        # Update status to processing
                        submission.status = SubmissionStatus.PROCESSING
                        await db.commit()
                        
                        # Queue task
                        grade_submission_task.apply_async(
                            args=[submission.id],
                            queue="grading",
                            priority=5  # Normal priority
                        )
                        queued_count += 1
                        logger.info(f"Queued submission {submission.id} for grading")
                    except Exception as e:
                        logger.error(f"Failed to queue submission {submission.id}: {e}")
                
                # Update exam status to completed
                exam.status = ExamStatus.COMPLETED
                await db.commit()
            
            logger.info(f"Processed {len(completed_exams)} exams, queued {queued_count} submissions")
            return {
                "processed_exams": len(completed_exams),
                "queued_submissions": queued_count
            }
            
        except Exception as e:
            logger.error(f"Error in process_completed_exams: {e}", exc_info=True)
            return {"error": str(e)}


@celery_app.task(name="app.tasks.grading.close_expired_sessions")
def close_expired_sessions():
    """
    Close ACTIVE sessions past their expiry and create empty submissions if missing.
    Runs every 5 minutes (configure in Celery beat).
    """
    logger.info("Starting scheduled task: close_expired_sessions")
    loop = asyncio.get_event_loop()
    return loop.run_until_complete(_close_expired_sessions_async())


async def _close_expired_sessions_async():
    async with AsyncSessionLocal() as db:
        import uuid
        try:
            now = datetime.utcnow()
            # Find active sessions past expiry
            result = await db.execute(
                select(ExamSession)
                .where(ExamSession.status == SessionStatus.ACTIVE)
                .options(
                    selectinload(ExamSession.submissions), 
                    selectinload(ExamSession.exam).selectinload(Exam.task_types)
                )
            )
            sessions = result.scalars().all()

            closed = 0
            created_empty = 0
            for session in sessions:
                hard_deadline = session.expires_at
                if session.exam and session.exam.end_time:
                    hard_deadline = min(hard_deadline, session.exam.end_time)
                if now >= hard_deadline:
                    session.status = SessionStatus.EXPIRED
                    session.submitted_at = session.submitted_at or hard_deadline
                    closed += 1
                    # Ensure a submission exists
                    submission = session.submissions[0] if session.submissions else None
                    if not submission:
                        # Compute total max score
                        exam = session.exam
                        total_max = 100.0
                        if exam and exam.task_types:
                            total_max = sum(task_type.max_score for task_type in exam.task_types)
                        submission = Submission(
                            id=str(uuid.uuid4()),
                            session_id=session.id,
                            student_id=session.student_id,
                            status=SubmissionStatus.UPLOADED,
                            max_score=total_max,
                            submitted_at=hard_deadline
                        )
                        db.add(submission)
                        created_empty += 1

            await db.commit()
            logger.info(f"Closed {closed} expired sessions, created {created_empty} empty submissions")
            return {"closed_sessions": closed, "created_empty_submissions": created_empty}
        except Exception as e:
            logger.error(f"Error in close_expired_sessions: {e}", exc_info=True)
            return {"error": str(e)}


@celery_app.task(name="app.tasks.grading.retry_failed_submissions")
def retry_failed_submissions():
    """
    Retry submissions that failed AI processing
    Runs every hour
    """
    logger.info("Starting scheduled task: retry_failed_submissions")
    
    loop = asyncio.get_event_loop()
    result = loop.run_until_complete(_retry_failed_submissions_async())
    
    return result


async def _retry_failed_submissions_async():
    """Retry failed submissions with limited retry count"""
    async with AsyncSessionLocal() as db:
        try:
            # Find flagged submissions with retries left
            result = await db.execute(
                select(Submission)
                .where(
                    Submission.status == SubmissionStatus.FLAGGED,
                    Submission.ai_retry_count < 3,
                    Submission.ai_error.isnot(None)
                )
            )
            failed_submissions = result.scalars().all()
            
            if not failed_submissions:
                logger.info("No failed submissions to retry")
                return {"retried_submissions": 0}
            
            logger.info(f"Found {len(failed_submissions)} failed submissions to retry")
            
            retried_count = 0
            for submission in failed_submissions:
                try:
                    # Reset status
                    submission.status = SubmissionStatus.PROCESSING
                    submission.ai_error = None
                    await db.commit()
                    
                    # Queue task with lower priority
                    grade_submission_task.apply_async(
                        args=[submission.id],
                        queue="grading",
                        priority=3  # Lower priority than new submissions
                    )
                    retried_count += 1
                    logger.info(f"Queued failed submission {submission.id} for retry")
                except Exception as e:
                    logger.error(f"Failed to retry submission {submission.id}: {e}")
            
            return {"retried_submissions": retried_count}
            
        except Exception as e:
            logger.error(f"Error in retry_failed_submissions: {e}", exc_info=True)
            return {"error": str(e)}


@celery_app.task(name="app.tasks.grading.cleanup_old_results")
def cleanup_old_results():
    """
    Cleanup old task results from Redis
    Runs daily at 3 AM
    """
    logger.info("Starting scheduled task: cleanup_old_results")
    
    try:
        # Celery automatically handles result expiration
        # This task is just a placeholder for any additional cleanup
        logger.info("Task results cleanup completed")
        return {"status": "completed"}
    except Exception as e:
        logger.error(f"Error in cleanup_old_results: {e}", exc_info=True)
        return {"error": str(e)}

File: src/tasks/grading.rs
use anyhow::{Context, Result};
use serde_json::{json, Value};
use sqlx::types::Json;
use sqlx::PgPool;
use time::{OffsetDateTime, PrimitiveDateTime};

use crate::core::state::AppState;
use crate::core::time::primitive_now_utc as now_primitive;
use crate::db::models::{Exam, ExamSession, Submission, SubmissionImage};
use crate::db::types::{ExamStatus, SessionStatus, SubmissionStatus};
use crate::services::ai_grading::{AiGradingService, GradeRequest};

pub(crate) async fn claim_next_submission(pool: &PgPool) -> Result<Option<String>> {
    let now = now_primitive();
    let id = sqlx::query_scalar::<_, String>(
        "WITH candidate AS (
            SELECT id FROM submissions
            WHERE status IN ($1, $2)
              AND ai_request_started_at IS NULL
            ORDER BY CASE WHEN status = $1 THEN 0 ELSE 1 END,
                     COALESCE(ai_retry_count, 0),
                     created_at
            FOR UPDATE SKIP LOCKED
            LIMIT 1
        )
        UPDATE submissions
        SET status = $3,
            ai_request_started_at = $4,
            ai_error = NULL
        FROM candidate
        WHERE submissions.id = candidate.id
        RETURNING submissions.id",
    )
    .bind(SubmissionStatus::Uploaded)
    .bind(SubmissionStatus::Processing)
    .bind(SubmissionStatus::Processing)
    .bind(now)
    .fetch_optional(pool)
    .await
    .context("Failed to claim submission")?;

    Ok(id)
}

pub(crate) async fn grade_submission(
    state: &AppState,
    ai: &AiGradingService,
    submission_id: &str,
) -> Result<()> {
    let submission =
        fetch_submission(state.db(), submission_id).await?.context("Submission not found")?;

    if !matches!(submission.status, SubmissionStatus::Processing | SubmissionStatus::Flagged) {
        tracing::info!(submission_id, status = ?submission.status, "Skipping grading");
        return Ok(());
    }

    let session =
        fetch_session(state.db(), &submission.session_id).await?.context("Session not found")?;
    let exam = fetch_exam(state.db(), &session.exam_id).await?.context("Exam not found")?;

    let mut images = fetch_images(state.db(), &submission.id).await?;
    images.sort_by_key(|image| image.order_index);

    if images.is_empty() {
        return flag_submission(
            state.db(),
            &submission.id,
            "No images available for AI grading",
            vec!["no_images".to_string()],
            false,
        )
        .await;
    }

    let storage = state.storage().ok_or_else(|| anyhow::anyhow!("S3 storage not configured"))?;

    let mut image_urls = Vec::new();
    for image in &images {
        if !image.file_path.starts_with("submissions/") {
            return flag_submission(
                state.db(),
                &submission.id,
                "Image is stored in local storage",
                vec!["storage_mismatch".to_string()],
                true,
            )
            .await;
        }

        let url = storage
            .presign_get(&image.file_path, std::time::Duration::from_secs(3600))
            .await
            .context("Failed to generate presigned URL")?;
        image_urls.push(url);
    }

    let (task_description, reference_solution, rubric, total_max_score) =
        build_task_prompt(state.db(), &exam, &session).await?;

    let request = GradeRequest {
        images: image_urls,
        task_description,
        reference_solution,
        rubric,
        max_score: total_max_score,
        chemistry_rules: None,
        submission_id: Some(submission.id.clone()),
    };

    let started_at = submission.ai_request_started_at.unwrap_or_else(now_primitive);
    let queue_latency =
        (started_at.assume_utc() - submission.created_at.assume_utc()).as_seconds_f64();

    let mut result = match ai.grade_submission(request).await {
        Ok(value) => value,
        Err(err) => {
            tracing::error!(submission_id, error = %err, "AI grading failed");
            metrics::counter!("grading_jobs_total", "status" => "failed").increment(1);
            return flag_submission(
                state.db(),
                &submission.id,
                &err.to_string(),
                vec!["ai_processing_error".to_string()],
                true,
            )
            .await;
        }
    };

    let unreadable = result.get("unreadable").and_then(|value| value.as_bool()).unwrap_or(false);

    let completed_at = now_primitive();
    let duration = (completed_at.assume_utc() - started_at.assume_utc()).as_seconds_f64();

    if unreadable {
        let reason = result
            .get("unreadable_reason")
            .and_then(|value| value.as_str())
            .unwrap_or("Изображение нечитаемо");
        metrics::counter!("grading_jobs_total", "status" => "unreadable").increment(1);
        return flag_submission(
            state.db(),
            &submission.id,
            reason,
            vec!["unreadable_images".to_string()],
            false,
        )
        .await;
    }

    if let Some(map) = result.as_object_mut() {
        map.remove("_metadata");
    }

    let total_score = result.get("total_score").and_then(|value| value.as_f64());

    let feedback =
        result.get("feedback").and_then(|value| value.as_str()).map(|value| value.to_string());

    sqlx::query(
        "UPDATE submissions SET status = $1, ai_score = $2, ai_analysis = $3, ai_comments = $4,
            ai_processed_at = $5, ai_request_completed_at = $6, ai_request_duration_seconds = $7,
            ai_error = NULL, is_flagged = FALSE, flag_reasons = $8, updated_at = $9
         WHERE id = $10",
    )
    .bind(SubmissionStatus::Preliminary)
    .bind(total_score)
    .bind(Json(result.clone()))
    .bind(feedback)
    .bind(completed_at)
    .bind(completed_at)
    .bind(duration)
    .bind(Json(Vec::<String>::new()))
    .bind(completed_at)
    .bind(&submission.id)
    .execute(state.db())
    .await
    .context("Failed to update submission")?;

    metrics::counter!("grading_jobs_total", "status" => "success").increment(1);
    metrics::histogram!("grading_duration_seconds").record(duration);
    metrics::histogram!("grading_queue_latency_seconds").record(queue_latency);

    if let Some(per_page) = result.get("per_page_transcriptions").and_then(|value| value.as_array())
    {
        for (idx, image) in images.iter().enumerate() {
            if let Some(text) = per_page.get(idx).and_then(|value| value.as_str()) {
                sqlx::query(
                    "UPDATE submission_images SET ocr_text = $1, processed_at = $2, is_processed = TRUE
                     WHERE id = $3",
                )
                .bind(text)
                .bind(completed_at)
                .bind(&image.id)
                .execute(state.db())
                .await
                .context("Failed to persist OCR transcription")?;
            }
        }
    }

    tracing::info!(submission_id, "AI grading succeeded");

    Ok(())
}

pub(crate) async fn process_completed_exams(state: &AppState) -> Result<()> {
    let now = now_primitive();

    let exams = sqlx::query_as::<_, Exam>(
        "SELECT id, title, description, start_time, end_time, duration_minutes, timezone,
                max_attempts, allow_breaks, break_duration_minutes, auto_save_interval,
                status, created_by, created_at, updated_at, published_at, settings
         FROM exams
         WHERE status IN ($1, $2)
           AND end_time <= $3",
    )
    .bind(ExamStatus::Active)
    .bind(ExamStatus::Published)
    .bind(now)
    .fetch_all(state.db())
    .await
    .context("Failed to fetch completed exams")?;

    if exams.is_empty() {
        return Ok(());
    }

    let mut queued = 0;

    for exam in &exams {
        let queued_ids = sqlx::query_scalar::<_, String>(
            "UPDATE submissions s
             SET status = $1,
                 ai_request_started_at = NULL,
                 updated_at = $2
             FROM exam_sessions es
             WHERE es.id = s.session_id
               AND es.exam_id = $3
               AND s.status = $4
               AND s.ai_request_started_at IS NULL
             RETURNING s.id",
        )
        .bind(SubmissionStatus::Processing)
        .bind(now)
        .bind(&exam.id)
        .bind(SubmissionStatus::Uploaded)
        .fetch_all(state.db())
        .await
        .context("Failed to queue submissions for processing")?;
        queued += queued_ids.len();

        sqlx::query("UPDATE exams SET status = $1, updated_at = $2 WHERE id = $3")
            .bind(ExamStatus::Completed)
            .bind(now)
            .bind(&exam.id)
            .execute(state.db())
            .await
            .context("Failed to mark exam as completed")?;
    }

    tracing::info!(
        processed_exams = exams.len(),
        queued_submissions = queued,
        "Processed completed exams"
    );
    metrics::counter!("exams_processed_total").increment(exams.len() as u64);
    metrics::counter!("submissions_queued_total").increment(queued as u64);

    Ok(())
}

pub(crate) async fn close_expired_sessions(state: &AppState) -> Result<()> {
    let now = OffsetDateTime::now_utc();

    #[derive(sqlx::FromRow)]
    struct SessionRow {
        id: String,
        expires_at: PrimitiveDateTime,
        exam_end_time: Option<PrimitiveDateTime>,
    }

    let sessions = sqlx::query_as::<_, SessionRow>(
        "SELECT s.id, s.expires_at, e.end_time AS exam_end_time
         FROM exam_sessions s
         LEFT JOIN exams e ON e.id = s.exam_id
         WHERE s.status = $1",
    )
    .bind(SessionStatus::Active)
    .fetch_all(state.db())
    .await
    .context("Failed to fetch active sessions")?;

    let mut closed = 0;

    for session in sessions {
        let hard_deadline = match session.exam_end_time {
            Some(end) => {
                if end < session.expires_at {
                    end
                } else {
                    session.expires_at
                }
            }
            None => session.expires_at,
        };

        if now.unix_timestamp() < hard_deadline.assume_utc().unix_timestamp() {
            continue;
        }

        sqlx::query(
            "UPDATE exam_sessions SET status = $1, submitted_at = COALESCE(submitted_at, $2), updated_at = $3 WHERE id = $4",
        )
        .bind(SessionStatus::Expired)
        .bind(hard_deadline)
        .bind(now_primitive())
        .bind(&session.id)
        .execute(state.db())
        .await
        .context("Failed to expire session")?;

        closed += 1;
    }

    tracing::info!(closed_sessions = closed, "Closed expired sessions");
    metrics::counter!("expired_sessions_closed_total").increment(closed as u64);

    Ok(())
}

pub(crate) async fn retry_failed_submissions(state: &AppState) -> Result<()> {
    let submissions = sqlx::query_scalar::<_, String>(
        "SELECT id FROM submissions
         WHERE status = $1
           AND COALESCE(ai_retry_count, 0) < 3
           AND ai_error IS NOT NULL",
    )
    .bind(SubmissionStatus::Flagged)
    .fetch_all(state.db())
    .await
    .context("Failed to fetch failed submissions")?;

    let mut retried = 0;
    let now = now_primitive();

    for submission_id in submissions {
        let updated = sqlx::query(
            "UPDATE submissions
             SET status = $1,
                 ai_retry_count = COALESCE(ai_retry_count, 0) + 1,
                 ai_error = NULL,
                 ai_request_started_at = NULL,
                 updated_at = $2
             WHERE id = $3",
        )
        .bind(SubmissionStatus::Processing)
        .bind(now)
        .bind(&submission_id)
        .execute(state.db())
        .await
        .context("Failed to requeue failed submission")?;

        if updated.rows_affected() > 0 {
            retried += 1;
        }
    }

    tracing::info!(retried_submissions = retried, "Retried failed submissions");
    metrics::counter!("submissions_retried_total").increment(retried as u64);

    Ok(())
}

async fn build_task_prompt(
    pool: &PgPool,
    exam: &Exam,
    session: &ExamSession,
) -> Result<(String, String, Value, f64)> {
    let task_types = crate::repositories::task_types::list_by_exam(pool, &exam.id)
        .await
        .context("Failed to fetch task types")?;

    let assignments = &session.variant_assignments.0;
    let mut descriptions = Vec::new();
    let mut reference_solutions = Vec::new();
    let mut rubric_items = Vec::new();
    let mut total_max_score = 0.0;

    for task_type in task_types {
        if let Some(variant_id) = assignments.get(&task_type.id) {
            let variant = crate::repositories::task_types::find_variant_by_id(pool, variant_id)
                .await
                .context("Failed to fetch variant")?;

            if let Some(variant) = variant {
                descriptions.push(format!(
                    "Задача {}: {}\n{}\n\nВариант:\n{}",
                    task_type.order_index + 1,
                    task_type.title,
                    task_type.description,
                    variant.content
                ));

                if let Some(reference) = variant.reference_solution {
                    reference_solutions.push(format!(
                        "Эталонное решение для задачи {}:\n{}",
                        task_type.order_index + 1,
                        reference
                    ));
                }

                rubric_items.push(json!({
                    "task_type": task_type.title,
                    "max_score": task_type.max_score,
                    "criteria": "Оценивать по критериям в системном промпте"
                }));

                total_max_score += task_type.max_score;
            }
        }
    }

    if total_max_score == 0.0 {
        total_max_score =
            exam.settings.0.get("max_score").and_then(|value| value.as_f64()).unwrap_or(100.0);
    }

    let rubric = json!({
        "criteria": rubric_items,
        "total_max_score": total_max_score,
    });

    let task_description = descriptions.join("\n\n");
    let reference_solution = if reference_solutions.is_empty() {
        "См. критерии оценивания".to_string()
    } else {
        reference_solutions.join("\n\n")
    };

    Ok((task_description, reference_solution, rubric, total_max_score))
}

async fn fetch_submission(pool: &PgPool, submission_id: &str) -> Result<Option<Submission>> {
    crate::repositories::submissions::find_by_id(pool, submission_id)
        .await
        .context("Failed to fetch submission")
}

async fn fetch_session(pool: &PgPool, session_id: &str) -> Result<Option<ExamSession>> {
    crate::repositories::sessions::find_by_id(pool, session_id)
        .await
        .context("Failed to fetch session")
}

async fn fetch_exam(pool: &PgPool, exam_id: &str) -> Result<Option<Exam>> {
    crate::repositories::exams::find_by_id(pool, exam_id).await.context("Failed to fetch exam")
}

async fn fetch_images(pool: &PgPool, submission_id: &str) -> Result<Vec<SubmissionImage>> {
    crate::repositories::images::list_by_submission(pool, submission_id)
        .await
        .context("Failed to fetch images")
}

async fn flag_submission(
    pool: &PgPool,
    submission_id: &str,
    reason: &str,
    flag_reasons: Vec<String>,
    increment_retry: bool,
) -> Result<()> {
    let now = now_primitive();
    let retry_expr =
        if increment_retry { "ai_retry_count = COALESCE(ai_retry_count,0) + 1," } else { "" };
    let query = format!(
        "UPDATE submissions SET status = $1, ai_error = $2, is_flagged = TRUE, flag_reasons = $3,
            ai_request_completed_at = $4, ai_request_duration_seconds = $5, {retry_expr} updated_at = $6
         WHERE id = $7",
    );

    sqlx::query(&query)
        .bind(SubmissionStatus::Flagged)
        .bind(reason)
        .bind(Json(flag_reasons))
        .bind(now)
        .bind(0.0)
        .bind(now)
        .bind(submission_id)
        .execute(pool)
        .await
        .context("Failed to flag submission")?;

    Ok(())
}
